spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/converter/PubSubMessageConverter.java

```diff
@@ -34,13 +34,13 @@
 	 * @param headers the headers of the message
 	 * @return the PubsubMessage ready to be sent
 	 */
-	PubsubMessage toMessage(Object payload, Map<String, String> headers);
+	PubsubMessage toPubSubMessage(Object payload, Map<String, String> headers);
 
 	/**
 	 * Convert the payload of a given {@code PubsubMessage} to a desired Java type.
 	 * @param message the message containing the payload of the object
 	 * @param payloadType the desired type of the object
 	 * @return the object converted from the message's payload
 	 */
-	<T> T fromMessage(PubsubMessage message, Class<T> payloadType);
+	<T> T fromPubSubMessage(PubsubMessage message, Class<T> payloadType);
```

Comment:
I think we need to revise this contract to avoid generics.
See the same `MessageConverter` for example:
```
Object fromMessage(Message<?> message, Class<?> targetClass);
```

So, we provide target class to expect, but that doesn't mean that return is going to be in exactly expected type.

The point is that the mentioned Jackson's default typing is really ignoring provided type and consults for that the provided type info in the JSON. Something similar matters with standard Java deserialization: that really does not have any effect what type you provide since I'm just going to deserialize from bytes.

Reply:
@artembilan Going back to this point, in the Jackson case, can't the client just call `fromPubSubMessage(message, Object.class)` or otherwise risk the `ClassCastException`?

Reply:
That's correct. That's exactly what happens in the `GenericJackson2JsonRedisSerializer` where I studied this approach from:
```
@Override
public Object deserialize(@Nullable byte[] source) throws SerializationException {
	return deserialize(source, Object.class);
}
```
And there is no any `ClassCastException`

Reply:
The casting would be done in our `JacksonPubSubMessageConverter`.

Reply:
I'm thinking of changing it back to how I had it before with the generics.

Reply:
Well, that's why I commented somewhere that there must not be any generics at all.
That's why  a `MessageConverter` has a contract like:
```
Object fromMessage(Message<?> message, Class<?> targetClass);
```
So, this really says us that even if we specify some type, that doesn't mean that a result object is going to be of that type.

Reply:
I understand, but what's wrong with actually requiring the result object to match the specified type?